airflow/models/dagrun.py

```diff
@@ -263,48 +262,34 @@ def update_state(self, session=None):
         Determines the overall state of the DagRun based on the state
         of its TaskInstances.
 
+        :param finished_tasks: The finished tasks of this run
+        :type finished_tasks: list[airflow.models.TaskInstance]
         :return: State
         """
 
         dag = self.get_dag()
-
-        tis = self.get_task_instances(session=session)
-        self.log.debug("Updating state for %s considering %s task(s)", self, len(tis))
+        ready_tis = []
+        tis = [ti for ti in self.get_task_instances(session=session) if ti.state != State.REMOVED]
```

Comment:
did u notice REMOVED task used to count towards SUCCESS dagrun? I raised https://github.com/apache/airflow/pull/6954

Reply:
do you have timings on performance benefit in a larger DAG say 500 tasks or 5000 tasks?

Reply:
No I didn't notice that, do you think it has something to do with this changes?

Reply:
I didn't try larger dags but I believe that the impact will be huge

Reply:
~What was your reason for filtering our removed TIs here?~ Oh we did this anyway in the next loop. Gotcha.

I think it would be better if we filtered it by using the existing `state` parameter of `get_task_instances` rather than having to filter it like this.

We could either do it by passing in a list of states and exclude REMOVED, or if we change `get_task_instances` to return a "Result" (i.e. don't do `return tis.all()` in there, but instead `return tis` -- this will behave almost the same when iterated over) we can then extend the filtering here. For example:

```suggestion
        tis = self.get_task_instances(session=session).fliter(TaskInstance.state != State.REMOVED).all()
```

Reply:
> do you think it has something to do with this chan

no, REMOVED state of TI leading to SUCCESS state of DagRun is existing bug, before your changes.